crates/hyperdrive-math/src/lp/remove.rs

```diff
@@ -1,6 +1,312 @@
+use ethers::types::{I256, U256};
+use eyre::{eyre, Result};
+use fixed_point::{fixed, FixedPoint};
+
+use super::math::SHARE_PROCEEDS_MAX_ITERATIONS;
 use crate::State;
 
-impl State {}
+impl State {
+    /// Allows an LP to burn shares and withdraw from the pool.
+    pub fn calculate_remove_liquidity(
+        &self,
+        current_block_timestamp: U256,
+        active_lp_total_supply: FixedPoint,
+        withdrawal_shares_total_supply: FixedPoint,
+        lp_shares: FixedPoint,
+        min_output_per_share: FixedPoint,
+        minimum_transaction_amount: FixedPoint,
+    ) -> Result<(FixedPoint, FixedPoint)> {
+        // Ensure that the amount of LP shares to remove is greater than or
+        // equal to the minimum transaction amount.
+        if lp_shares < minimum_transaction_amount {
+            return Err(eyre!("Minimum transaction amount not met"));
+        }
+
+        // Get vault share price and proceed without checkpoint.
+        let vault_share_price = self.vault_share_price();
+
+        let mut state = self.clone();
+        // Burn the LP's shares.
+        state.info.lp_total_supply -= lp_shares.into();
+        let active_lp_total_supply = active_lp_total_supply - lp_shares;
+
+        // Mint an equivalent amount of withdrawal shares.
+        let withdrawal_shares_total_supply = withdrawal_shares_total_supply + lp_shares;
```

Comment:
you can declare the variable as `mut` in the fn signature and then use `+=`. I am not sure which is better, tbh, but the latter pattern is what we use e.g. in [yieldspace.rs](https://github.com/delvtech/hyperdrive-rs/blob/dpaiton/calc-open-short/crates/hyperdrive-math/src/yield_space.rs#L274) and [short/open.rs](https://github.com/delvtech/hyperdrive-rs/blob/dpaiton/calc-open-short/crates/hyperdrive-math/src/short/open.rs#L43).

Reply:
It's been my preference to not use `mut` when possible and just redeclare.  Avoiding mutability is kinda Rust's bag.

Reply:
oh ok, I didn't know that. Should we change those examples I gave you then?

Reply:
I wouldn't go out of your way but in general we should avoid mutability unless needed for a for-loop scenario etc.